src/core/Akka.Remote/Transport/Streaming/StreamAssociationHandle.cs

```diff
@@ -0,0 +1,238 @@
+using System;
+using System.IO;
+using System.Threading.Tasks;
+using Akka.Actor;
+using Google.ProtocolBuffers;
+
+namespace Akka.Remote.Transport.Streaming
+{
+    public class StreamAssociationHandle : AssociationHandle
+    {
+        private readonly Stream _stream;
+        private readonly AsyncQueue<ByteString> _writeQueue;
+        private IHandleEventListener _eventListener;
+
+        private readonly byte[] _readBuffer;
+        private readonly byte[] _writeBuffer;
+        private readonly TimeSpan _flushWaitOnShutdown;
+
+        private readonly TaskCompletionSource<bool> _initialized;
+        private readonly TaskCompletionSource<bool> _stopped;
+        private readonly Task _writeLoop;
+
+        /// <summary>
+        /// Returns true if gracefully stopped, otherwise false.
+        /// </summary>
+        public Task<bool> Stopped => _stopped.Task;
+
+        public StreamAssociationHandle(StreamTransportSettings settings, Stream stream, Address localAddress, Address remoteAddress)
+            : base(localAddress, remoteAddress)
+        {
+            _stream = stream;
+            _writeQueue = new AsyncQueue<ByteString>();
+
+            _readBuffer = new byte[settings.StreamReadBufferSize];
+            _writeBuffer = new byte[settings.StreamWriteBufferSize];
+            _flushWaitOnShutdown = settings.FlushWaitTimeout;
+
+            _initialized = new TaskCompletionSource<bool>();
+            _stopped = new TaskCompletionSource<bool>();
+            _writeLoop = Task.Run(() => WriteLoop());
+        }
+
+        internal void Initialize(IHandleEventListener eventListener)
```

Comment:
So correct me if I'm wrong - but does each instance of this transport use two separate threads perpetually? One for writing and another for reading respectively?

Or do the `WriteAsync` / `ReadAsync` calls use an I/O completion port and release those threads back until data is available / flushed? I'm assuming the latter but just want to be certain.